Properly handle all Monitoring exporters being disabled

pickypg opened the pull request:
When monitoring exporters are all disabled, which must be done explicitly, _and_ monitoring collection is enabled, then any call to `_xpack/monitoring/_bulk` will create a task that never closes _and_ ES collection will stop happening because a semaphore is never marked as completed.

This also simplifies the async `ExportBulk` code by removing the third step (second async step, `close`) entirely because it was entirely unnecessary by both implementations.

Closes #40919

pickypg committed:
Handle Monitoring Exporters all Disabled

When monitoring exporters are all disabled, which must be done
explicitly, _and_ monitoring collection is enabled, then
any call to `_xpack/monitoring/_bulk` will create a task that
never closes _and_ ES collection will stop happening because
a semaphore is never marked as completed.

This also simplifies the async `ExportBulk` code by removing the
third step (second async step, `close`) entirely because it was
entirely unnecessary by both implementations.

elasticmachine commented:
Pinging @elastic/es-core-features

pickypg committed:
simplify no exporters test

pickypg commented:
Ping @elastic/stack-monitoring

ycombinator reviewed (approved):
Functionally, this PR LGTM. I tested it as follows:

1. I enabled Monitoring collection and created a local exporter that was disabled:

   ```
   PUT _cluster/settings
   {
     "persistent": {
       "xpack.monitoring.exporters": {
         "my_remote": {
           "type": "local",
           "enabled": false
         }
       },
       "xpack.monitoring.collection.enabled": true
     }
   }
   ```

2. Then I started up Kibana so it would make requests to the `POST _monitoring/bulk` API every 10s.

3. I checked the list of running tasks with `GET _cat/tasks` and verified that there were no `cluster:admin/xpack/monitoring/bulk` tasks piling up without getting closed.

As for a code review, I'd really prefer someone from the @elastic/es-core-features team to take a look, maybe @jakelandis since he has been in the vicinity of ES monitoring code relatively recently?

jakelandis commented on x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/local/LocalBulk.java:
```diff
@@ -52,13 +52,10 @@
+    protected void doAdd(Collection<MonitoringDoc> docs) throws ExportException {
         ExportException exception = null;
 
         for (MonitoringDoc doc : docs) {
```
Do we still need a isClosed() check here ?  (now against the AtomicBoolean)

pickypg replied:
We do not. I removed code that cheated and called the implementations directly, which makes this whole method guarded by `add`. Also, the whole `add-flush` flow was always synchronous so we never really made use of the inner-loop escape.

jakelandis commented on x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/Exporters.java:
```diff
@@ -225,7 +233,7 @@ private void doExport(final ExportBulk bulk, final Collection<MonitoringDoc> doc
         } catch (ExportException e) {
             exceptionRef.set(e);
         } finally {
-            bulk.close(lifecycleState() == Lifecycle.State.STARTED, ActionListener.wrap(r -> {
```
there are alot of states here, so i am kinda getting lost... but i _think_ this different then before since this can now flush if the Lifecycle.State.STARTED == false. Not sure if this an issue

pickypg replied:
You're correct. This is a subtle change in behavior, but it was always an unnecessary one. The clients used by the exporter implementations will be closed when the process shuts down even if we're in the middle of a flush / close, and the `exporter` don't exist until the lifecycle is started, which means we couldn't have even gotten here anyway.

jakelandis commented:
Overall, I think this looks good. However, it is pretty hard to separate the refactoring from the issue this addresses and the refactorings are kinda hard to follow if they are semantically equivalent (at least for me). 

I left a couple questions in the review.

pickypg commented:
> it is pretty hard to separate the refactoring from the issue this addresses and the refactorings are kinda hard to follow if they are semantically equivalent (at least for me).

Sorry about that. I had also been confused when I revisited the old `close` / `doClose` code to the point that I wondered if the bug was in that code. It was after making that change and further debugging (after I dropped that code and updated tests to handle it) that I discovered the core change ( https://github.com/elastic/elasticsearch/pull/40920/files#diff-a8c42761f9a33243a3256b61442318a6R185 ) that fixes the bug.

I could split these out if you'd prefer, but it sounds like you've got a handle on it now.

jakelandis reviewed (approved):
LGTM thanks for the additional explanations.

pickypg commented:
7.x: https://github.com/elastic/elasticsearch/commit/64a05e522f1911aa8cb38a2f6c4fb018ff0850fa (https://github.com/elastic/elasticsearch/pull/41043)
7.0: https://github.com/elastic/elasticsearch/commit/42ed969a022f4d58b946ee6105704e8d838c580f (https://github.com/elastic/elasticsearch/pull/41044)
6.7: https://github.com/elastic/elasticsearch/commit/1226a1f2fd05a92e19cd3a0a59e1370e06d47438 (https://github.com/elastic/elasticsearch/pull/41045)
